micro-rdk/src/common/data_manager.rs

```diff
@@ -262,43 +252,139 @@ where
+                        let data = match data {
+                            Ok(data) => data,
+                            Err(err) => {
+                                log::error!(
+                                    "error decoding readings for collector key ({:?}): {:?}",
+                                    collector_key,
+                                    err
+                                );
+                                break false;
+                            }
+                        };
+                        let upload_request = DataCaptureUploadRequest {
+                            metadata: Some(UploadMetadata {
+                                part_id: self.part_id.clone(),
+                                component_type: collector_key.component_type.clone(),
+                                r#type: DataType::TabularSensor.into(),
+                                component_name: collector_key.r_name.clone(),
+                                method_name: collector_key.method.to_string(),
+                                ..Default::default()
+                            }),
+                            sensor_contents: data,
+                        };
+                        match app_client
+                            .upload_data(upload_request)
+                            .or(async {
+                                async_io::Timer::after(Duration::from_millis(1000)).await;
+                                Err(AppClientError::AppClientRequestTimeout)
+                            })
+                            .await
+                        {
+                            Ok(_) => {
+                                current_chunk_size = next_message.len();
+                                current_chunk.push(next_message);
+                                break true;
+                            }
+
+                            // If the request takes too long to elicit a response, we don't know whether
+                            // the upload was successful on app's side. We've decided that we'd rather
+                            // potentially lose some data than upload duplicate data, and so we opt to
+                            // assume the best and consume the messages
+                            Err(AppClientError::AppClientRequestTimeout) => {
+                                current_chunk_size = next_message.len();
+                                current_chunk.push(next_message);
+                                break true;
+                            }
+                            Err(err) => return Err(err),
+                        };
+                    } else {
+                        current_chunk_size += next_message.len();
+                        current_chunk.push(next_message);
+                    }
                 };
-                app_client.upload_data(upload_request).await?;
+
+                // all of the data in the current chunk has been successfully uploaded, so we
```

Comment:
This is outside the inner loop, so it isn't actually flushing the current chunk, it is all chunks. It feels related to my above concern about the Nth chunk.

Reply:
As above.